proposals/2024-01-26_classic-histograms-stored-as-native-histograms.md

```diff
@@ -0,0 +1,296 @@
+* Once you are done, switch off scraping classic histograms (or drop via metric relabel as needed).
+
+Use case: custom histograms feature is off, but exponential histograms (current native histograms feature) is on.
+
+* For histograms that have no exponential histogram version, the same applies as above.
+* For histograms that have exponential histogram version, nothing happens, exponential histograms will be kept.
+
+### Documentation
+
+* Feature should be marked as experimental, it depends on another experimental feature, namely native histograms.
+* Configuration option to be documented.
+* Impact on queries and PromQL functions to be specified.
+
+### Data structures
+
+New constant to define the schema number to mean custom buckets (e.g. 127 ?).
+
+Remote write protocol
+
+The `message.Histogram` type to be expanded with nullable repeated custom_buckets field that list the custom bucket definitions (except `+Inf`, which is implicit). There should be a comment which specifies which schema number means that we need to even look at this field. It should be a validation error to find this field null if the custom bucket schema number is used.
+
+Internal representation
+
+The `histogram.Histogram` and `histogram.FloatHistogram` types to get an additional field that has the custom bucket definitions or reference id for interned custom bucket definition. Doesn’t matter as it should be accessible via getter interface. (We’d have to see how the PromQL engine uses the bucket definitions to nail down what exact interface to put here.)
+
+Chunk representation
+
+We propose that the current integer histogram and float histogram chunk is expanded with custom bucket field encoding (and type is encoded in schema). Leaving exact format to PR author.
+
+Chunk iterator
+
+No change to interface. The raw chunk iterator can load the custom bucket definitions once and reuse that (like it does with counter reset hint header).
+
+## Open questions
+
+None.
+
+## Answered questions
+
+* Would we ever want to store the old representation and the new one at the same time?
+  *Answer:* YES. Already should work via the `existing scrape_classic_histograms` option.
+* What to do in queries if custom histogram and exponential histogram meet or customer histogram and float sample?
+  *Answer:* same as today with float vs native histogram, that is calculate the result if it makes mathematical sense. For example multiplying a custom histogram with the number 2.0 makes sense. In case of histograms they need to rescaled to match their schema.
```

Comment:
"rescaled" is confusing (as we also use it if we multiply a histogram with a float). Maybe "change resolution" or something. It should also be added that this might not work precisely, and that in that case we rather give a warning and don't do the calculation instead of doing something wonky.

Examples (not necessarily to add here, just to create understanding for the reviewers):

* Custom histogram with boundaries 1, 2, 4, 8 and no sample in the +Inf bucket → can be merged with any exponential histogram.
* Anything in the +Inf bucket: Doesn't work with exponential histogram.
* Custom histogram with boundaries 1, 5, 10, 20, +Inf added to another custom histogram with 1, 10, 20, +Inf: We can do the math by first merging the 5-bucket and the 10-bucket together in the first histogram and then add both, resulting in a 1, 10, 20 histogram.